File: bindings/c/src/gvector2d.cpp
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

#include <nux/gvector2d.h>
#include "handles.hpp"

NUXC_EXTERNC
nux_gvector2d_t*
nux_gvector2d_create (
    float x,
    float y
) {
    return new nux_gvector2d_t{nux::gvector2d<float>{x, y}};
}

NUXC_EXTERNC
void
nux_gvector2d_destroy (
    nux_gvector2d_t* gvector2d
) {
    delete gvector2d;
}

NUXC_EXTERNC
float
nux_gvector2d_get_x (
    nux_gvector2d_t* gvector2d
) {
    return gvector2d->handle.x;
}

NUXC_EXTERNC
void
nux_gvector2d_set_x (
    nux_gvector2d_t* gvector2d,
    float new_x
) {
    gvector2d->handle.x = new_x;
}

NUXC_EXTERNC
float
nux_gvector2d_get_y (
    nux_gvector2d_t* gvector2d
) {
    return gvector2d->handle.y;
}

NUXC_EXTERNC
void
nux_gvector2d_set_y (
    nux_gvector2d_t* gvector2d,
    float new_y
) {
    gvector2d->handle.y = new_y;
}

File: bindings/swift/Nux/Sources/Nux/GVector2D.swift
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

import Foundation
import CNux

public class GVector2D {
    public init (
        x: Float,
        y: Float
    ) {
        self.handle = nux_gvector2d_create(x, y)
    }

    deinit {
        nux_gvector2d_destroy(self.handle)
    }

    public var x: Float {
        get {
            return nux_gvector2d_get_x(self.handle)
        }
        set {
            nux_gvector2d_set_x(self.handle, newValue)
        }
    }

    public var y: Float {
        get {
            return nux_gvector2d_get_y(self.handle)
        }
        set {
            nux_gvector2d_set_y(self.handle, newValue)
        }
    }

    public var handle: OpaquePointer? = nil
}
